lib/msal-common/src/cache/CacheManager.ts

```diff
@@ -826,6 +1009,88 @@ export abstract class CacheManager implements ICacheManager {
         await Promise.all(removedCredentials);
     }
 
+    /**
+     * Migrates a single-tenant account and all it's associated alternate cross-tenant account objects in the
+     * cache into a condensed multi-tenant account object with tenant profiles.
+     * @param accountKey
+     * @param accountEntity
+     * @param logger
+     * @returns
+     */
+    protected updateOutdatedCachedAccount(
+        accountKey: string,
+        accountEntity: AccountEntity | null,
+        logger?: Logger
+    ): AccountEntity | null {
+        // Only update if account entity is defined and has no tenantProfiles object (is outdated)
+        if (accountEntity && accountEntity.isSingleTenant()) {
+            this.commonLogger?.verbose(
+                "updateOutdatedCachedAccount: Found a single-tenant (outdated) account entity in the cache, migrating to multi-tenant account entity"
+            );
+
+            // Get keys of all accounts belonging to user
+            const matchingAccountKeys = this.getAccountKeys().filter(
+                (key: string) => {
+                    return key.startsWith(accountEntity.homeAccountId);
+                }
+            );
+
+            // Get all account entities belonging to user
+            const accountsToMerge: AccountEntity[] = [];
+            matchingAccountKeys.forEach((key: string) => {
+                const account = this.getCachedAccountEntity(key);
+                if (account) {
+                    accountsToMerge.push(account);
+                }
+            });
+
+            // Set base account to home account if available, any account if not
+            const baseAccount =
+                accountsToMerge.find((account) => {
+                    return tenantIdMatchesHomeTenant(
+                        account.realm,
+                        account.homeAccountId
+                    );
+                }) || accountsToMerge[0];
+
+            // Populate tenant profiles built from each account entity belonging to the user
+            baseAccount.tenantProfiles = accountsToMerge.map(
+                (account: AccountEntity) => {
+                    return {
+                        tenantId: account.realm,
+                        localAccountId: account.localAccountId,
+                        name: account.name,
+                        isHomeTenant: tenantIdMatchesHomeTenant(
+                            account.realm,
+                            account.homeAccountId
+                        ),
+                    };
+                }
+            );
+
+            const updatedAccount = Object.assign(new AccountEntity(), {
```

Comment:
I thought we had a helper for this `CacheManager.toObject`. Not super critical since I plan to refactor AccountEntity into a type after this is merged

Reply:
Right, these were used interchangeably, and I think they're functionally equivalent. Not sure why we have CacheManager.toObject, but since this is in fact inside CacheManager I'll use that for now for consistency.